### ISSUE TITLE
Problem with (de)serialization unions created with `GenerateOneOf`

### ISSUE THREAD

**Issue by dpraimeyuu** [mcintyre321/OneOf#167]
`GenerateOneOf` is a huge boost when it comes to avoiding both boilerplate and complaints from fellow devs that "omg implicit ugliness" 🚀 

Before version `3.x.x` (I assume no Source Gen at all), `OneOfBase` had a parameterless constructor that enabled (de)serialization and led to a tiny package [OneOf.Serialization](https://github.com/dpraimeyuu/OneOf.Serialization/).

Now, all generated `OneOfBase`s have no parameterless constructor which makes the following not possible to achieve:
```cs
[JsonConverter(typeof(OneOfJsonConverter<Status>))]
public class Status : OneOfBase<Idle, Running, Completed> {
	[JsonConstructor]
    private Status() { }

    public static implicit operator Status(Idle value) => value == null? null : new Status(value);
    public Status(Idle idle) : base(0, idle) {}
    public static implicit operator Status(Running value) => value == null? null : new Status(value);
	public Status(Running running) : base(1, null, idle) {}
    public static implicit operator Status(Completed value) => value == null? null : new Status(value);
	public Status(Completed completed) : base(2, null, null, completed)
}
```

I tried to make some acrobatics with `GetConstructors` (yup, Reflection...) to bypass no possibility to get `OneOfBase` default constructor but didn't progress that much.

I would like to combine power of both  `GenerateOneOf` and `OneOfJsonConverter` to have great `OneOf` experience but have no idea what's the best way of handling that.

Initially, I thought of suggesting to add a default constructor to every generated `OneOfBase` but I am not sure how it fits the "roadmap" and the library vision.

Mentioning three core contributors to get attention @mcintyre321 @romfir @zspitz 😅

**Comment by mcintyre321** [comment]
NewtonSoft apparently can handle that using:


JsonSerializerSettings settings = new JsonSerializerSettings
{
    ConstructorHandling = ConstructorHandling.AllowNonPublicDefaultConstructor
};


Would that work for you?

**Comment by dpraimeyuu** [comment]
Unfortunately, it won't work as "DU" new type inherits from `OneOfBase<A, B, C>` which does not allow a default constructor.

I took a [break](https://www.youtube.com/watch?v=f84n5oFoZBc&ab_channel=ClojureTV) and after that, I had a simple idea - I was able to comply with the requirement coming from deriving from `OneOfBase<A, B, C>` by putting three case-driven constructors:
```cs
[GenerateOneOf]
[JsonConverter(typeof(OneOfJsonConverter<Status>))]
public class Status : OneOfBase<Idle, Running, Completed> {
	public Status(Idle idle) : base(idle) {}
	public Status(Running running) : base(running) {}
	public Status(Completed completed) : base(completed) {}
}
```

Looks pretty legit 😅 could be even source generated on "my" end.

Closing the issue, thanks for your response, Harry! 🚀
